### PR TITLE
Common: Custom Chain File Integration

### PR DESCRIPTION
**Pull request by holgerd77**
This PR will be adding functionality for a more convenient and flexible way to integrate custom chains into Common instances,  will push a more complete code base here throughout the day.

This is an early push since I also improved some typings along the way on functions like `hardfork()` or `consensusType()` and I was not sure if this might have side effects, so I wanted to run the additional tests from CI.

### CODE REVIEW

**Comment by holgerd77** [comment]
Ok, this is now ready for review. 😄 

To add on the missing explanation: this PR introduces a more flexible way to add custom chains with a new optional `customChains` constructor option where an arbitrary number of chains can be passed in. This should allow for a range of new use cases, e.g. if people want to build some library/tools integrating some of our libraries or just Common and supporting a range of their own custom chains (or just: some other chains than the ones we support).

The PR furthermore improves on the overall custom chain situation and combines all the custom chain tests into a dedicated test file and comes with a revised README section. Hope this all sets us better up for custom chain support.

One special thing: I added a new (still optional, backwards compatibility) parameter `defaultHardfork` to the chain files. This allows - e.g. on custom chains - to define by-chain default HFs and is especially useful for chains which just don't include the currently defined default HF, this would otherwise be a constant source of error with initialization throwing if people not explicitly pass an existing HF for their custom chain. For now the solution is fully backwards compatible, pointing to `istanbul` for all existing supported chains. We might want to leverage this ourselves in the future since this can allow us to solve situations which lately occurred with `muirGlacier` where we needed to stick to `istanbul` on all chains as default while the PoW chains could have moved on to `muirGlacier`. With the new option we have at least the possibility for a more fine grained approach if we want to. 😄

**Comment by jochem-brouwer** [comment]
Will give this a review in a minute. Just a comment here:

> where we needed to stick to istanbul on all chains as default while the PoW chains could have moved on to `muirGlacier`

This only applies to ETH mainnet as this delays the difficulty bomb by a certain amount and is thus dependent on the height of a chain. (I'm not sure if ETC has something similar)

File: packages/common/README.md
Comment (jochem-brouwer) [inline packages/common/README.md:135]: From this description I am not entirely sure what this does. I can initialize a common which is set to the default `mainnet`, but I can also pass an array with extra chain parameters - how do I activate them? Should I activate them by setting it as hardfork?

File: packages/common/src/chains/index.ts
Comment (jochem-brouwer) [inline packages/common/src/chains/index.ts:34]: Hmm. So custom chains are treated on a different foot than "normal" chains. Can we either move these custom chains into the special field of `chains.names` or disallow the custom chain name `names`? (Very unlikely to happen, but it will probably mess up the internal logic if someone uses this name).

File: packages/common/src/index.ts
Comment (jochem-brouwer) [inline packages/common/src/index.ts:108]: Ok, so I can only get a custom chain by name, not by ID, right?

File: packages/common/src/index.ts
Comment (jochem-brouwer) [inline packages/common/src/index.ts:96]: Here `customChains` is an argument - can we not use `this.customChains` and remove the argument?

File: packages/common/src/types.ts
Comment (jochem-brouwer) [inline packages/common/src/types.ts:20]: I was about to suggest this to indeed make `defaultHardfork` mandatory.

File: packages/common/tests/data/energyWebChain.json
Comment (jochem-brouwer) [inline packages/common/tests/data/energyWebChain.json:2]: Awesome test data 😄

File: packages/common/tests/customChains.ts
Comment (jochem-brouwer) [inline packages/common/tests/customChains.ts:83]: Ah OK so this is how it works - could this be added in the README?

**Review by jochem-brouwer — COMMENTED** [review]
Some questions, but looks great in general 😄

File: packages/common/src/chains/mainnet.json
Comment (holgerd77) [reply packages/common/src/chains/mainnet.json:5]: No, this would be breaking if we change the default HF along the road. We would need to wait until the next major release for VM and Common.

File: packages/common/README.md
Comment (holgerd77) [reply packages/common/README.md:107]: Updated (next push).

File: packages/common/README.md
Comment (holgerd77) [reply packages/common/README.md:135]: Have rewritten this, hope this makes it more clear.

File: packages/common/src/chains/index.ts
Comment (holgerd77) [reply packages/common/src/chains/index.ts:34]: Are you maybe misreading the code here (or I get your message wrong)? Custom chains are just normally added to the `names` dict here and the other entries are not affected in any way.

Generally I would want to move away from this `chains/index.ts` file altogether - therefore the deprecation notes. I would already want to move this to logic to the main `index.ts` file but then I kept it like this also for some backwards compatibility considerations.

File: packages/common/src/index.ts
Comment (holgerd77) [reply packages/common/src/index.ts:96]: No, since the method is static and custom chains from the instance are not available yet. This is a private method though, so should be unproblematic.

File: packages/common/src/index.ts
Comment (holgerd77) [reply packages/common/src/index.ts:108]: No, that's the same logic as before here - `chainParams` -> `initializedChains` is just a renaming. The code for returning from a number is some lines above and still works for baked-in as well as custom chains. In fact baked-in chains and custom chains are treated exactly the same throughout the whole library I would assume (if I am not missing something).

File: packages/common/src/types.ts
Comment (holgerd77) [reply packages/common/src/types.ts:20]: If someone is using this type this could otherwise break things unfortunately.

File: packages/common/tests/customChains.ts
Comment (holgerd77) [reply packages/common/tests/customChains.ts:83]: Have added.

File: packages/common/tests/data/energyWebChain.json
Comment (holgerd77) [reply packages/common/tests/data/energyWebChain.json:2]: This is taken from #968 😄 , I would actually want to wait for a quick ok from @matextrem the author until we merge here.

**Comment by holgerd77** [comment]
Ok, did the changes and rebased.

**Comment by jochem-brouwer** [comment]
Added a `blocked` label to await @matextrem's approval for the testdata.

File: packages/common/src/chains/index.ts
Comment (jochem-brouwer) [reply packages/common/src/chains/index.ts:34]: At line 30 it sets `chains[name]` to `chain` - so if you'd use the `name` "names", then it'd first set `chains.names` to the chain and then later overwrite it. Not sure if that causes problems?

File: packages/common/src/index.ts
Comment (jochem-brouwer) [reply packages/common/src/index.ts:96]: Ah - woops - did not see that this was a `static` method.

File: packages/common/src/index.ts
Comment (jochem-brouwer) [reply packages/common/src/index.ts:108]: Yup you are right, I misread this code.

File: packages/common/src/types.ts
Comment (jochem-brouwer) [reply packages/common/src/types.ts:20]: Yep but we can do this in the next major release (maybe add this to the V6 tracking issue?)

**Review by jochem-brouwer — DISMISSED** [review]
Will approve but lets await @matextrem's approval for the test data.

File: packages/common/README.md
Comment (cgewecke) [inline packages/common/README.md:120]:
```suggestion
If you just want to change certain parameters on a chain configuration it can also be convenient to use 
```

**Review by cgewecke — DISMISSED** [review]
LGTM!

**Comment by holgerd77** [comment]
@matextrem Is it ok if we use your examples here? I would otherwise push another commit tomorrow and remove again.

**Comment by holgerd77** [comment]
Ok, I now replaced the test files here with some more neutral onces. This would need another approval for merge.

**Review by cgewecke — APPROVED** [review]
Test file changes LGTM. 

Re-approving 🙂

**Comment by ryanio** [comment]
ah this didn't make it: https://github.com/ethereumjs/ethereumjs-monorepo/pull/1034#discussion_r553517646 would be a good typo to fix

**Comment by holgerd77** [comment]
@ryanio have added to #905

**Pull request merged**
